base/reduce.jl

```diff
@@ -36,7 +36,7 @@ mul_prod(x::Real, y::Real)::Real = x * y
 
 ## foldl && mapfoldl
 
-function mapfoldl_impl(f, op, nt::NamedTuple{(:init,)}, itr, i...)
+function mapfoldl_impl(f::F, op::OP, nt::NamedTuple{(:init,)}, itr, i...) where {F,OP}
```

Comment:
Migth be worth a comment, as the `where {F,OP}` is here purely for this unusual reason of forcing specialization, which may not be evident to the next person touching these lines (who might as well, with good intentions, revert this PR unknowingly :) )

Reply:
We might want to do that broadly across Julia's code base. Grep string in https://github.com/JuliaLang/julia/issues/32761#issuecomment-517855527

Reply:
Ah right! Would it stretch too much the meaning of `@specialize` to extend this macro to formalize this technique? E.g. `@specialize(f), @specialize(op)` in this case, even if no `@nospecialize` was active?

Reply:
I think it would be great if `@specialize` did this. I didn't even realize we had it! There's a discourse post from a while back making this exact suggestion.